services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/api_client/_base.py

```diff
@@ -0,0 +1,141 @@
+                    try:
+                        logger.debug("Calling retry_on_errors")
+                        response: Response = await request_func(zelf, *args, **kwargs)
+                        return response
+                    except ConnectError as e:
+                        assert e._request
+                        logger.info(
+                            "[%s/%s]Retry. Unexpected ConnectError while requesting '%s %s': %s",
+                            attempt.retry_state.attempt_number,
+                            zelf._request_max_retries,
+                            e._request.method,
+                            e._request.url,
+                            f"{e=}",
+                        )
+                        raise _RetryRequestError(e) from e
+                    except TransportError as e:
+                        raise ClientTransportError from e
+        except _RetryRequestError as e:
+            # raise original exception
+            assert e.__cause__
+            raise e.__cause__
+
+    return request_wrapper
+
+
+def expect_status(expected_code: int):
+    """
+    NOTE: always apply after `retry_on_errors`
+
+    raises:
+    - `UnexpectedStatusError`
+    - `httpx.HTTPError`
+    """
+
+    def decorator(
+        request_func: Callable[..., Awaitable[Response]]
+    ) -> Callable[..., Awaitable[Response]]:
+        assert asyncio.iscoroutinefunction(request_func)
+
+        @functools.wraps(request_func)
+        async def request_wrapper(zelf: "BaseHThinClient", *args, **kwargs) -> Response:
+            logger.debug("Calling expect_status")
+
+            response = await request_func(zelf, *args, **kwargs)
+            if response.status_code != expected_code:
+                raise UnexpectedStatusError(response, expected_code)
+
+            return response
+
+        return request_wrapper
+
+    return decorator
+
+
+class BaseHThinClient:
+    SKIP_METHODS: set[str] = {"close"}
+
+    def __init__(
+        self,
+        *,
+        request_max_retries: int,
+        base_url: Optional[URLTypes] = None,
+        timeout: Optional[TimeoutTypes] = None,
+    ) -> None:
+        self._request_max_retries: int = request_max_retries
+
+        client_args: dict[str, Any] = {}
+        if base_url:
+            client_args["base_url"] = base_url
+        if timeout:
+            client_args["timeout"] = timeout
+        self._client: AsyncClient = AsyncClient(**client_args)
+
+        # ensure all user defined public methods return `httpx.Response``
+
+        pubic_methods = [
+            x
+            for x in inspect.getmembers(self, predicate=inspect.ismethod)
+            if not (x[0].startswith("_") or x[0] in self.SKIP_METHODS)
+        ]
+
+        for _, method in pubic_methods:
+            signature = inspect.signature(method)
+            if signature.return_annotation != Response:
+                raise WrongReturnType(method, signature.return_annotation)
```

Comment:
This is actually static-analysis that should be performed before runtime (e.g. via mypi/pylance etc)

Reply:
Ideally this should run at import time! But I'm not sure how to achieve it. Will add a note for the future.

Reply:
I might have not managed to express the idea behind this check correctly.

I'd like for users to now be able to define a method, in their subclass, which returns something different from `httpx.Response`. In that case an error should be raised.

I don't see how `mypy` would check for this. Someone could add a method which is not decorated and returns a string. Nothing will catch this case (which is what I want to check).

As I already said, Ideally the check should run at import time, but for now it's done when instantiating the method.